otelcol/buffered_core.go

```diff
@@ -0,0 +1,71 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// This logger implements zapcore.Core and is based on zaptest/observer.
+
+package otelcol // import "go.opentelemetry.io/collector/otelcol"
+
+import (
+	"sync"
+
+	"go.uber.org/zap/zapcore"
+)
+
+type loggedEntry struct {
+	zapcore.Entry
+	Context []zapcore.Field
+}
+
+func newBufferedCore(enab zapcore.LevelEnabler) *bufferedCore {
+	return &bufferedCore{LevelEnabler: enab}
+}
+
+var _ zapcore.Core = (*bufferedCore)(nil)
+
+type bufferedCore struct {
+	zapcore.LevelEnabler
+	mu      sync.RWMutex
+	logs    []loggedEntry
+	context []zapcore.Field
+}
+
+func (bc *bufferedCore) Level() zapcore.Level {
+	return zapcore.LevelOf(bc.LevelEnabler)
+}
+
+func (bc *bufferedCore) Check(ent zapcore.Entry, ce *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	if bc.Enabled(ent.Level) {
+		return ce.AddCore(ent, bc)
+	}
+	return ce
+}
+
+func (bc *bufferedCore) With(fields []zapcore.Field) zapcore.Core {
+	return &bufferedCore{
+		LevelEnabler: bc.LevelEnabler,
+		logs:         bc.logs,
+		context:      append(bc.context, fields...),
+	}
+}
+
+func (bc *bufferedCore) Write(ent zapcore.Entry, fields []zapcore.Field) error {
+	all := make([]zapcore.Field, 0, len(fields)+len(bc.context))
+	all = append(all, bc.context...)
+	all = append(all, fields...)
+	bc.mu.Lock()
+	bc.logs = append(bc.logs, loggedEntry{ent, all})
+	bc.mu.Unlock()
+	return nil
+}
+
+func (bc *bufferedCore) Sync() error {
+	return nil
+}
+
+func (bc *bufferedCore) TakeLogs() []loggedEntry {
+	bc.mu.Lock()
+	logs := bc.logs
+	bc.logs = nil
```

Comment:
Should we make it so that core is unusable after calling `TakeLogs` once? Right now, if misused we can lose some logs (e.g. if you were to first call `TakeLogs` and then replace the core, calls to `Write` in this time slice would be lost forever).

If we make it so that `Write` returns an error after the first call to `TakeLogs` we can at least see that something went wrong (or, alternatively, make it so that the API makes it impossible for you to call things in the wrong order, although that would mean some coupling between `bufferedCore` and `collectorCore`)

Reply:
I think this is a good restriction to start with.  If we need to later we could open it up, but based on our current use case this enforcement seems reasonable.